README.md
# #DeployDB

**Who, what, where and when.**

[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/lookout/deploydb)
[![Build Status](https://travis-ci.org/lookout/deploydb.svg?branch=master)](https://travis-ci.org/lookout/deploydb)

**To learn more please visit the [DeployDB website](http://hackers.lookout.com/deploydb/)**

DeployDB is a tool to provide a single source of truth for artifact-based
deployments through multiple environments. It is intended to fit within an
existing infrastructure where CI and deployment orchestration are already
provided by other tools (e.g. Jenkins, Rundeck).



## Helpful links

 * [Gitter chat room for DeployDB](https://gitter.im/lookout/deploydb)
 * [DeployDB homepage](http://hackers.lookout.com/deploydb/)
 * [Hacking DeployDB](http://hackers.lookout.com/deploydb/hacking.html)
 * [REST API documentation](http://hackers.lookout.com/deploydb/restapi/)
 * [Groovydocs](http://hackers.lookout.com/deploydb/groovydoc)
 * [Internal design](http://hackers.lookout.com/deploydb/internals.html)
 * [Run the application](http://hackers.lookout.com/deploydb/running.html)


CHANGELOG.md
# DeployDB Changelog

## 0.2.0 - Antares

* [#3](https://github.com/lookout/deploydb/issues/3) - Model what "Promotion Criteria" looks like
* [#9](https://github.com/lookout/deploydb/issues/9) - Allow defining a Pipeline via YAML
* [#14](https://github.com/lookout/deploydb/issues/14) - DeployDB needs some persistent queueing abstraction
* [#26](https://github.com/lookout/deploydb/issues/26) - Invalid JSON in cucumber tests doesn't properly raise exceptions
* [#32](https://github.com/lookout/deploydb/issues/32) - Fix the request artifact with support for pageNumber=0 and pageSize=0
* [#35](https://github.com/lookout/deploydb/issues/35) - Administrators should be able to configure webhooks for deployment creation
* [#44](https://github.com/lookout/deploydb/issues/44) - DeployDB should be able to reload configuration with an API call
* [#77](https://github.com/lookout/deploydb/issues/77) - 201 CREATED responses should include LOCATION header of newly created resource which can be GET'd later.
* [#92](https://github.com/lookout/deploydb/issues/92) - Upgrade to Dropwizard 0.8.0 final
* [#95](https://github.com/lookout/deploydb/issues/95) - Add REST API to GET Results of a promotion associated with the deployment
* [#99](https://github.com/lookout/deploydb/issues/99) - Add restart logic and DB persistence
* [#102](https://github.com/lookout/deploydb/pull/102) - Upgrade to Dropwizard 0.8 final
* [#103](https://github.com/lookout/deploydb/issues/103) - When used with MySql on AWS, Some SQL commands failed to execute 
* [#104](https://github.com/lookout/deploydb/pull/104) - Added instructions for how to run the deployDb application
* [#106](https://github.com/lookout/deploydb/issues/106) - Webhooks should have a `type` field to differentiate event types
* [#107](https://github.com/lookout/deploydb/issues/107) - Add deploydb support for whoas configuration
* [#110](https://github.com/lookout/deploydb/issues/110) - When multiple webhooks are configured, only one webhook is invoked
* [#111](https://github.com/lookout/deploydb/issues/111) - Webhook invocation fails when environment webhook has empty event type like started
* [#113](https://github.com/lookout/deploydb/pull/113) - Fix the multiple webhooks invocation use case
* [#114](https://github.com/lookout/deploydb/pull/114) - Using flyway java(groovy) migrations for sql compatibility with diff DBs
* [#115](https://github.com/lookout/deploydb/pull/115) - Issue 111 webhook invocation fails for for empty started config
* [#116](https://github.com/lookout/deploydb/pull/116) - Upgrade to Dropwizard 0.8 final
* [#119](https://github.com/lookout/deploydb/pull/119) - Issue 106 webhook content type for events
* [#120](https://github.com/lookout/deploydb/issues/120) - Define new REST API to reload configuration with an API call
* [#121](https://github.com/lookout/deploydb/pull/121) - Defined REST API for configuration reload
* [#122](https://github.com/lookout/deploydb/pull/122) - Added configReload admin task to reload the config, if no deployments are in progress
* [#123](https://github.com/lookout/deploydb/pull/123) - Fix V2,4-6 migrations to work against MySQL
* [#124](https://github.com/lookout/deploydb/pull/124) - Remove integer type lengths in schema migrations


## 0.1.0 - Minimum Loveable Pipeline

* [#27](https://github.com/lookout/deploydb/pull/27) - Fix for issue #23
* [#28](https://github.com/lookout/deploydb/issues/28) - Decouple docs generation from test tasks
* [#29](https://github.com/lookout/deploydb/issues/29) - The request should fail when creating an Artifact without a version
* [#33](https://github.com/lookout/deploydb/issues/33) - In the create artifact request, validate max length for group, name, version & sourceUrl
* [#36](https://github.com/lookout/deploydb/issues/36) - We need some ability to deserialize YAML into Models
* [#37](https://github.com/lookout/deploydb/pull/37) - Submitting some outstanding changes I had lying around
* [#40](https://github.com/lookout/deploydb/pull/40) - Stop generating groovy and asciidoctor docs when running tests
* [#43](https://github.com/lookout/deploydb/issues/43) - Expose Services via a read API
* [#45](https://github.com/lookout/deploydb/issues/45) - How DeployDb loads yaml-based/configuration-driven models

changelog.rb
#!/usr/bin/env ruby

require 'json'
require 'net/https'
require 'uri'

def uri_for(project, milestone)
  return URI("https://api.github.com/repos/lookout/#{project}/issues?state=closed&milestone=#{milestone}")
end

def changelog_for(project, milestone)
  uri = uri_for(project, milestone)
  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = true
  response = http.get(uri.request_uri)

  raise "Status #{response.code}" unless response.code.to_i == 200

  body = JSON.parse(response.body)

  body.reverse.each do |issue|
    puts "* [##{issue['number']}](#{issue['html_url']}) - #{issue['title']}"
  end
end

print 'What milestone? > '
milestone = STDIN.gets.chomp
project = 'deploydb'

puts
puts "Computing changelog for '#{project}' and milestone '#{milestone}'"
puts
puts '-------------------'

changelog_for(project, milestone)


features/step_definitions/basic_http_steps.rb
#
# NOTE: Many of these steps were shamelessly borrowed from the
# lookout-rack-test project which can be found here:
# <https://github.com/lookout/lookout-rack-test>

When /^I (GET|PUT|POST|PATCH|DELETE)( to)? "(.*?)"$/ do |verb, unused, url_path|
  verb = verb.downcase.to_sym
  send(verb, render_string(url_path))
end

When /^I (GET|PUT|POST|PATCH|DELETE)( to)? "([^"]*)" with:$/ do |verb, unused, url, params|
  options = {}

  params.hashes.each do |row|
    options[row[:Name]] = render_string(row[:Value])
  end

  verb = verb.downcase.to_sym
  send(verb, render_string(url), options)
end

When /^I (GET|PUT|POST|PATCH|DELETE) to "(.*?)" with the JSON:$/ do |verb, url, body|
  verb = verb.downcase.to_sym
  send(verb, render_string(url), render_string(body))
end

Then /^the response (should be|is) (\d+)$/ do |verb, code|
  expect(last_response.status).to eq(code), "expected #{code} got #{last_response.status} \n\n #{last_response.body}
"
end

Then /^I should be redirected to "([^"]+)"$/ do |path|
  expect(last_response.location).to end_with path
end

Then(/^the response body should include:$/) do |contents|
  expect(last_response.body).to match(contents)
end


Then /^I should receive the JSON:$/ do |json|
  template_vars['timestamp'] = Time.now.utc.iso8601
  expected = JSON.parse(render_string(json))

  begin
    received = JSON.parse(last_response.body)
  rescue JSON::ParserError => e
    raise "Unexpected response: #{last_response.body}\n#{e}"
  end

  expect(received).to eql(expected)
end




restapi/report.js
$(document).ready(function() {var formatter = new CucumberHTML.DOMFormatter($('.cucumber-report'));formatter.uri("api/artifact/creating.feature");
formatter.feature({
  "id": "artifact-create-apis",
  "description": "\nAs a RESTful client or user\nI should be able to create new artifacts in the system",
  "name": "Artifact CREATE APIs",
  "keyword": "Feature",
  "line": 1
});
formatter.before({
  "duration": 2661590000,
  "status": "passed"
});
formatter.before({
  "duration": 4839000,
  "status": "passed"
});
formatter.scenario({
  "id": "artifact-create-apis;creating-a-new-artifact",
  "tags": [
    {
      "name": "@freezetime",
      "line": 7
    }
  ],
  "description": "\nThe creation of an artifact technically only requires group, name and\nversion. However it\u0027s a good idea to always provide a `sourceUrl` as well\nin order future users of DeployDB to be able to backtrack an artifact all\nthe way to the binary itself",
  "name": "Creating a new artifact",
  "keyword": "Scenario",
  "line": 8,
  "type": "scenario"
});
formatter.step({
  "name": "I PUT to \"/api/v1/artifacts\" with:",
  "keyword": "When ",
  "line": 15,
  "doc_string": {
    "value": "  {\n    \"group\" : \"com.example.cucumber\",\n    \"name\" : \"cukes\",\n    \"version\" : \"1.2.345\",\n    \"sourceUrl\" : \"http://example.com/cucumber.jar\"\n  }",
    "line": 16,
    "content_type": ""
  }
});
formatter.step({
  "name": "the response should be 201",
  "keyword": "Then ",
  "line": 24
});
formatter.step({
  "name": "the body should be JSON:",
  "keyword": "And ",
  "line": 25,
  "doc_string": {
    "value": "  {\n    \"id\" : 1,\n    \"group\" : \"com.example.cucumber\",\n    \"name\" : \"cukes\",\n    \"version\" : \"1.2.345\",\n    \"sourceUrl\" : \"http://example.com/cucumber.jar\",\n    \"createdAt\" : \"{{created_timestamp}}\"\n  }",
    "line": 26,
    "content_type": ""
  }
});
formatter.match({
  "arguments": [
    {
      "val": "/api/v1/artifacts",
      "offset": 10
    }
  ],
  "location": "HttpSteps.groovy:24"
});
formatter.result({
  "duration": 612297000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "201",
      "offset": 23
    }
  ],
  "location": "HttpSteps.groovy:33"
});
formatter.result({
  "duration": 5804000,
  "status": "passed"
});
formatter.match({
  "location": "HttpSteps.groovy:41"
});
formatter.result({
  "duration": 64164000,
  "status": "passed"
});
formatter.after({
  "duration": 18321000,
  "status": "passed"
});
formatter.after({
  "duration": 488000,
  "status": "passed"
});
formatter.uri("api/artifact/deleting.feature");
formatter.feature({
  "id": "artifact-delete-apis",
  "description": "\nAs a RESTful client or user\nI should be *not* be able to delete artifacts",
  "name": "Artifact DELETE APIs",
  "keyword": "Feature",
  "line": 1
});
formatter.before({
  "duration": 363370000,
  "status": "passed"
});
formatter.scenario({
  "id": "artifact-delete-apis;attempting-to-delete-an-artifact",
  "description": "\nOnce artifacts are in DeployDB they\u0027re there forever. FOREVER.",
  "name": "Attempting to delete an artifact",
  "keyword": "Scenario",
  "line": 7,
  "type": "scenario"
});
formatter.step({
  "name": "there is an artifact",
  "keyword": "Given ",
  "line": 11
});
formatter.step({
  "name": "I DELETE \"/api/v1/artifacts/1\"",
  "keyword": "When ",
  "line": 12
});
formatter.step({
  "name": "the response should be 405",
  "keyword": "Then ",
  "line": 13
});
formatter.match({
  "location": "ArtifactSteps.groovy:8"
});
formatter.result({
  "duration": 31425000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "/api/v1/artifacts/1",
      "offset": 10
    }
  ],
  "location": "HttpSteps.groovy:20"
});
formatter.result({
  "duration": 36164000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "405",
      "offset": 23
    }
  ],
  "location": "HttpSteps.groovy:33"
});
formatter.result({
  "duration": 174000,
  "status": "passed"
});
formatter.after({
  "duration": 10582000,
  "status": "passed"
});
formatter.uri("api/artifact/reading.feature");
formatter.feature({
  "id": "artifact-read-apis",
  "description": "\nAs a RESTful client or user\nI should be able to read information about artifacts",
  "name": "Artifact READ APIs",
  "keyword": "Feature",
  "line": 1
});
formatter.before({
  "duration": 313276000,
  "status": "passed"
});
formatter.before({
  "duration": 58000,
  "status": "passed"
});
formatter.scenario({
  "id": "artifact-read-apis;fetching-an-artifact-by-id-that-exists",
  "tags": [
    {
      "name": "@freezetime",
      "line": 7
    }
  ],
  "description": "",
  "name": "Fetching an artifact by ID that exists",
  "keyword": "Scenario",
  "line": 8,
  "type": "scenario"
});
formatter.step({
  "name": "there is an artifact",
  "keyword": "Given ",
  "line": 10
});
formatter.step({
  "name": "I GET \"/api/v1/artifacts/1\"",
  "keyword": "When ",
  "line": 11
});
formatter.step({
  "name": "the response should be 200",
  "keyword": "Then ",
  "line": 12
});
formatter.step({
  "name": "the body should be JSON:",
  "keyword": "And ",
  "line": 13,
  "doc_string": {
    "value": "  {\n    \"id\" : 1,\n    \"group\" : \"com.example.cucumber\",\n    \"name\" : \"cucumber-artifact\",\n    \"version\" : \"1.0.1\",\n    \"sourceUrl\" : \"http://example.com/maven/com.example.cucumber/cucumber-artifact/1.0.1/cucumber-artifact-1.0.1.jar\",\n    \"createdAt\" : \"{{created_timestamp}}\"\n  }",
    "line": 14,
    "content_type": ""
  }
});
formatter.match({
  "location": "ArtifactSteps.groovy:8"
});
formatter.result({
  "duration": 5124000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "/api/v1/artifacts/1",
      "offset": 7
    }
  ],
  "location": "HttpSteps.groovy:16"
});
formatter.result({
  "duration": 55350000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "200",
      "offset": 23
    }
  ],
  "location": "HttpSteps.groovy:33"
});
formatter.result({
  "duration": 128000,
  "status": "passed"
});
formatter.match({
  "location": "HttpSteps.groovy:41"
});
formatter.result({
  "duration": 1739000,
  "status": "passed"
});
formatter.after({
  "duration": 7922000,
  "status": "passed"
});
formatter.after({
  "duration": 33000,
  "status": "passed"
});
formatter.before({
  "duration": 393520000,
  "status": "passed"
});
formatter.scenario({
  "id": "artifact-read-apis;fetching-an-artifact-by-id-that-doesn\u0027t-exist",
  "description": "",
  "name": "Fetching an artifact by ID that doesn\u0027t exist",
  "keyword": "Scenario",
  "line": 25,
  "type": "scenario"
});
formatter.step({
  "name": "I GET \"/api/v1/artifacts/1\"",
  "keyword": "When ",
  "line": 27
});
formatter.step({
  "name": "the response should be 404",
  "keyword": "Then ",
  "line": 28
});
formatter.match({
  "arguments": [
    {
      "val": "/api/v1/artifacts/1",
      "offset": 7
    }
  ],
  "location": "HttpSteps.groovy:16"
});
formatter.result({
  "duration": 52168000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "404",
      "offset": 23
    }
  ],
  "location": "HttpSteps.groovy:33"
});
formatter.result({
  "duration": 138000,
  "status": "passed"
});
formatter.after({
  "duration": 7835000,
  "status": "passed"
});
formatter.before({
  "duration": 401724000,
  "status": "passed"
});
formatter.scenario({
  "id": "artifact-read-apis;fetching-an-artifact-by-a-non-numeric-id",
  "tags": [
    {
      "name": "@error",
      "line": 30
    }
  ],
  "description": "",
  "name": "Fetching an artifact by a non-numeric ID",
  "keyword": "Scenario",
  "line": 31,
  "type": "scenario"
});
formatter.step({
  "name": "I GET \"/api/v1/artifacts/0xdeadbeef\"",
  "keyword": "When ",
  "line": 33
});
formatter.step({
  "name": "the response should be 400",
  "keyword": "Then ",
  "line": 34
});
formatter.match({
  "arguments": [
    {
      "val": "/api/v1/artifacts/0xdeadbeef",
      "offset": 7
    }
  ],
  "location": "HttpSteps.groovy:16"
});
formatter.result({
  "duration": 45295000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "400",
      "offset": 23
    }
  ],
  "location": "HttpSteps.groovy:33"
});
formatter.result({
  "duration": 178000,
  "status": "passed"
});
formatter.after({
  "duration": 8711000,
  "status": "passed"
});
formatter.before({
  "duration": 257788000,
  "status": "passed"
});
formatter.before({
  "duration": 53000,
  "status": "passed"
});
formatter.scenario({
  "id": "artifact-read-apis;fetching-an-artifact-by-the-(group,-name)-tuple",
  "tags": [
    {
      "name": "@freezetime",
      "line": 37
    }
  ],
  "description": "",
  "name": "Fetching an artifact by the (group, name) tuple",
  "keyword": "Scenario",
  "line": 38,
  "type": "scenario"
});
formatter.step({
  "name": "there is an artifact",
  "keyword": "Given ",
  "line": 40
});
formatter.step({
  "name": "I GET \"/api/v1/artifacts/by-module/com.example.cucumber:cucumber-artifact\"",
  "keyword": "When ",
  "line": 41
});
formatter.step({
  "name": "the response should be 200",
  "keyword": "Then ",
  "line": 42
});
formatter.step({
  "name": "the body should be JSON:",
  "keyword": "And ",
  "line": 43,
  "doc_string": {
    "value": "  {\n    \"id\" : 1,\n    \"group\" : \"com.example.cucumber\",\n    \"name\" : \"cucumber-artifact\",\n    \"version\" : \"1.0.1\",\n    \"sourceUrl\" : \"http://example.com/maven/com.example.cucumber/cucumber-artifact/1.0.1/cucumber-artifact-1.0.1.jar\",\n    \"createdAt\" : \"{{created_timestamp}}\"\n  }",
    "line": 44,
    "content_type": ""
  }
});
formatter.match({
  "location": "ArtifactSteps.groovy:8"
});
formatter.result({
  "duration": 2433000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "/api/v1/artifacts/by-module/com.example.cucumber:cucumber-artifact",
      "offset": 7
    }
  ],
  "location": "HttpSteps.groovy:16"
});
formatter.result({
  "duration": 57327000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "200",
      "offset": 23
    }
  ],
  "location": "HttpSteps.groovy:33"
});
formatter.result({
  "duration": 152000,
  "status": "passed"
});
formatter.match({
  "location": "HttpSteps.groovy:41"
});
formatter.result({
  "duration": 1486000,
  "status": "passed"
});
formatter.after({
  "duration": 6156000,
  "status": "passed"
});
formatter.after({
  "duration": 23000,
  "status": "passed"
});
formatter.before({
  "duration": 264967000,
  "status": "passed"
});
formatter.scenario({
  "id": "artifact-read-apis;fetching-an-artifact-by-the-(group,-name)-tuple-that-doesn\u0027t-exist",
  "description": "",
  "name": "Fetching an artifact by the (group, name) tuple that doesn\u0027t exist",
  "keyword": "Scenario",
  "line": 55,
  "type": "scenario"
});
formatter.step({
  "name": "I GET \"/api/v1/artifacts/by-module/com.example.cucumber:cucumber-artifact\"",
  "keyword": "When ",
  "line": 57
});
formatter.step({
  "name": "the response should be 404",
  "keyword": "Then ",
  "line": 58
});
formatter.match({
  "arguments": [
    {
      "val": "/api/v1/artifacts/by-module/com.example.cucumber:cucumber-artifact",
      "offset": 7
    }
  ],
  "location": "HttpSteps.groovy:16"
});
formatter.result({
  "duration": 41365000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "404",
      "offset": 23
    }
  ],
  "location": "HttpSteps.groovy:33"
});
formatter.result({
  "duration": 155000,
  "status": "passed"
});
formatter.after({
  "duration": 6051000,
  "status": "passed"
});
formatter.uri("api/artifact/updating.feature");
formatter.feature({
  "id": "artifact-update-apis",
  "description": "\nAs a RESTful client or user\nI should be *not* be able to update artifacts",
  "name": "Artifact UPDATE APIs",
  "keyword": "Feature",
  "line": 1
});
formatter.before({
  "duration": 298597000,
  "status": "passed"
});
formatter.scenario({
  "id": "artifact-update-apis;attempting-to-update-an-artifact",
  "description": "\nArtifacts must be considered immutable to ensure consistency of data from\nDeployDB. We can\u0027t different binaries with the same version, that\u0027s silly.",
  "name": "Attempting to update an artifact",
  "keyword": "Scenario",
  "line": 7,
  "type": "scenario"
});
formatter.step({
  "name": "there is an artifact",
  "keyword": "Given ",
  "line": 12
});
formatter.step({
  "name": "I PATCH \"/api/v1/artifacts/1\" with:",
  "keyword": "When ",
  "line": 13,
  "doc_string": {
    "value": "  {\n    \"version\" : \"9.0.0.0\"\n  }",
    "line": 14,
    "content_type": ""
  }
});
formatter.step({
  "name": "the response should be 405",
  "keyword": "Then ",
  "line": 19
});
formatter.match({
  "location": "ArtifactSteps.groovy:8"
});
formatter.result({
  "duration": 1750000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "/api/v1/artifacts/1",
      "offset": 9
    }
  ],
  "location": "HttpSteps.groovy:28"
});
formatter.result({
  "duration": 26598000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "405",
      "offset": 23
    }
  ],
  "location": "HttpSteps.groovy:33"
});
formatter.result({
  "duration": 104000,
  "status": "passed"
});
formatter.after({
  "duration": 14401000,
  "status": "passed"
});
formatter.uri("api/deployment/creating.feature");
formatter.feature({
  "id": "deployment-create-apis",
  "description": "\nAs a RESTful client or user\nI should be able to create new deployments in the system",
  "name": "Deployment CREATE APIs",
  "keyword": "Feature",
  "line": 1
});
formatter.before({
  "duration": 210472000,
  "status": "passed"
});
formatter.before({
  "duration": 44000,
  "status": "passed"
});
formatter.scenario({
  "id": "deployment-create-apis;creating-a-new-deployment",
  "tags": [
    {
      "name": "@freezetime",
      "line": 7
    }
  ],
  "description": "\nCreating a deployment assumes that we\u0027ve already got an Artifact stored\ninside the DeployDB database",
  "name": "Creating a new deployment",
  "keyword": "Scenario",
  "line": 8,
  "type": "scenario"
});
formatter.step({
  "name": "there is an artifact",
  "keyword": "Given ",
  "line": 13
});
formatter.step({
  "name": "I PUT to \"/api/v1/deployments\" with:",
  "keyword": "When ",
  "line": 14,
  "doc_string": {
    "value": "  {\n    \"artifactId\" : 1,\n    \"environment\" : \"pre-production\"\n  }",
    "line": 15,
    "content_type": ""
  }
});
formatter.step({
  "name": "the response should be 201",
  "keyword": "Then ",
  "line": 21
});
formatter.step({
  "name": "the body should be JSON:",
  "keyword": "And ",
  "line": 22,
  "doc_string": {
    "value": "  {\n    \"id\" : 1,\n    \"artifact\" : {\n      \"id\" : 1,\n      \"group\" : \"com.example.cucumber\",\n      \"name\" : \"cucumber-artifact\",\n      \"version\" : \"1.0.1\",\n      \"sourceUrl\" : \"http://example.com/maven/com.example.cucumber/cucumber-artifact/1.0.1/cucumber-artifact-1.0.1.jar\",\n      \"createdAt\" : \"{{created_timestamp}}\"\n    },\n    \"environment\" : \"pre-production\",\n    \"status\" : \"STARTED\",\n    \"createdAt\" : \"{{created_timestamp}}\"\n  }",
    "line": 23,
    "content_type": ""
  }
});
formatter.match({
  "location": "ArtifactSteps.groovy:8"
});
formatter.result({
  "duration": 1696000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "/api/v1/deployments",
      "offset": 10
    }
  ],
  "location": "HttpSteps.groovy:24"
});
formatter.result({
  "duration": 34503000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "201",
      "offset": 23
    }
  ],
  "location": "HttpSteps.groovy:33"
});
formatter.result({
  "duration": 106000,
  "status": "passed"
});
formatter.match({
  "location": "HttpSteps.groovy:41"
});
formatter.result({
  "duration": 1388000,
  "status": "passed"
});
formatter.after({
  "duration": 5091000,
  "status": "passed"
});
formatter.after({
  "duration": 19000,
  "status": "passed"
});
formatter.uri("api/deployment/reading.feature");
formatter.feature({
  "id": "deployment-read-apis",
  "description": "\nAs a RESTful client or user\nI should be able to read information about deployments",
  "name": "Deployment READ APIs",
  "keyword": "Feature",
  "line": 1
});
formatter.before({
  "duration": 266427000,
  "status": "passed"
});
formatter.before({
  "duration": 60000,
  "status": "passed"
});
formatter.scenario({
  "id": "deployment-read-apis;fetching-a-deployment-by-id",
  "tags": [
    {
      "name": "@freezetime",
      "line": 7
    }
  ],
  "description": "",
  "name": "Fetching a deployment by ID",
  "keyword": "Scenario",
  "line": 8,
  "type": "scenario"
});
formatter.step({
  "name": "there is a deployment",
  "keyword": "Given ",
  "line": 10
});
formatter.step({
  "name": "I GET \"/api/v1/deployments/1\"",
  "keyword": "When ",
  "line": 11
});
formatter.step({
  "name": "the response should be 200",
  "keyword": "Then ",
  "line": 12
});
formatter.step({
  "name": "the body should be JSON:",
  "keyword": "And ",
  "line": 13,
  "doc_string": {
    "value": "  {\n    \"id\" : 1,\n    \"artifact\" : {\n      \"id\" : 1,\n      \"group\" : \"com.example.cucumber\",\n      \"name\" : \"cucumber-artifact\",\n      \"version\" : \"1.0.1\",\n      \"sourceUrl\" : null,\n      \"createdAt\" : \"{{created_timestamp}}\"\n    },\n    \"environment\" : \"pre-production\",\n    \"status\" : \"STARTED\",\n    \"createdAt\" : \"{{created_timestamp}}\"\n  }",
    "line": 14,
    "content_type": ""
  }
});
formatter.match({
  "location": "DeploymentSteps.groovy:8"
});
formatter.result({
  "duration": 4706000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "/api/v1/deployments/1",
      "offset": 7
    }
  ],
  "location": "HttpSteps.groovy:16"
});
formatter.result({
  "duration": 39941000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "200",
      "offset": 23
    }
  ],
  "location": "HttpSteps.groovy:33"
});
formatter.result({
  "duration": 103000,
  "status": "passed"
});
formatter.match({
  "location": "HttpSteps.groovy:41"
});
formatter.result({
  "duration": 1984000,
  "status": "passed"
});
formatter.after({
  "duration": 5401000,
  "status": "passed"
});
formatter.after({
  "duration": 28000,
  "status": "passed"
});
formatter.before({
  "duration": 283708000,
  "status": "passed"
});
formatter.scenario({
  "id": "deployment-read-apis;fetching-a-deployment-by-id-that-doesn\u0027t-exist",
  "description": "",
  "name": "Fetching a deployment by ID that doesn\u0027t exist",
  "keyword": "Scenario",
  "line": 31,
  "type": "scenario"
});
formatter.step({
  "name": "I GET \"/api/v1/deployments/1\"",
  "keyword": "When ",
  "line": 33
});
formatter.step({
  "name": "the response should be 404",
  "keyword": "Then ",
  "line": 34
});
formatter.match({
  "arguments": [
    {
      "val": "/api/v1/deployments/1",
      "offset": 7
    }
  ],
  "location": "HttpSteps.groovy:16"
});
formatter.result({
  "duration": 31910000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "404",
      "offset": 23
    }
  ],
  "location": "HttpSteps.groovy:33"
});
formatter.result({
  "duration": 162000,
  "status": "passed"
});
formatter.after({
  "duration": 6094000,
  "status": "passed"
});
formatter.uri("api/deployment/updating.feature");
formatter.feature({
  "id": "deployment-update-apis",
  "description": "\nAs a RESTful client or user\nI should be able to update existing deployments in the system",
  "name": "Deployment UPDATE APIs",
  "keyword": "Feature",
  "line": 1
});
formatter.before({
  "duration": 268577000,
  "status": "passed"
});
formatter.before({
  "duration": 64000,
  "status": "passed"
});
formatter.scenario({
  "id": "deployment-update-apis;updating-a-deployment-that-doesn\u0027t-exist",
  "tags": [
    {
      "name": "@freezetime",
      "line": 37
    },
    {
      "name": "@error",
      "line": 37
    }
  ],
  "description": "",
  "name": "Updating a deployment that doesn\u0027t exist",
  "keyword": "Scenario",
  "line": 38,
  "type": "scenario"
});
formatter.step({
  "name": "I PATCH \"/api/v1/deployments/1\" with:",
  "keyword": "When ",
  "line": 40,
  "doc_string": {
    "value": "  {\n    \"status\" : \"FAILED\"\n  }",
    "line": 41,
    "content_type": ""
  }
});
formatter.step({
  "name": "the response should be 404",
  "keyword": "Then ",
  "line": 46
});
formatter.match({
  "arguments": [
    {
      "val": "/api/v1/deployments/1",
      "offset": 9
    }
  ],
  "location": "HttpSteps.groovy:28"
});
formatter.result({
  "duration": 29964000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "404",
      "offset": 23
    }
  ],
  "location": "HttpSteps.groovy:33"
});
formatter.result({
  "duration": 132000,
  "status": "passed"
});
formatter.after({
  "duration": 5389000,
  "status": "passed"
});
formatter.after({
  "duration": 26000,
  "status": "passed"
});
formatter.uri("api/health.feature");
formatter.feature({
  "id": "the-health-endpoint",
  "description": "\nAs a DeployDB administrator\nI should be able to view the health of the application",
  "name": "The health endpoint",
  "keyword": "Feature",
  "line": 1
});
formatter.before({
  "duration": 257210000,
  "status": "passed"
});
formatter.scenario({
  "id": "the-health-endpoint;fetching-health-under-normal-circumstances",
  "description": "\nAssuming the application is up and running properly, the built-in\nhealthcheck endpoint should return information about all the configured\nhealthchecks currently running inside of DeployDB",
  "name": "Fetching health under normal circumstances",
  "keyword": "Scenario",
  "line": 6,
  "type": "scenario"
});
formatter.step({
  "name": "I GET \"/healthcheck\" from the admin app",
  "keyword": "When ",
  "line": 12
});
formatter.step({
  "name": "the response should be 200",
  "keyword": "Then ",
  "line": 13
});
formatter.step({
  "name": "the body should be JSON:",
  "keyword": "And ",
  "line": 14,
  "doc_string": {
    "value": "{\n  \"deadlocks\" : {\n    \"healthy\" : true\n  },\n  \"sanity\" : {\n    \"healthy\" : true\n  },\n  \"hibernate\" : {\n    \"healthy\" : true\n  },\n  \"webhook\" : {\n    \"healthy\" : true\n  }\n}",
    "line": 15,
    "content_type": ""
  }
});
formatter.match({
  "arguments": [
    {
      "val": "/healthcheck",
      "offset": 7
    }
  ],
  "location": "HttpSteps.groovy:12"
});
formatter.result({
  "duration": 57082000,
  "status": "passed"
});
formatter.match({
  "arguments": [
    {
      "val": "200",
      "offset": 23
    }
  ],
  "location": "HttpSteps.groovy:33"
});
formatter.result({
  "duration": 160000,
  "status": "passed"
});
formatter.match({
  "location": "HttpSteps.groovy:41"
});
formatter.result({
  "duration": 1726000,
  "status": "passed"
});
formatter.after({
  "duration": 4951000,
  "status": "passed"
});
});